## LORA for text classification

In [ ]:
import pandas as pd, numpy as np

from peft import (
    LoraConfig, 
    get_peft_model, 
    TaskType,
    PeftModel
)

from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    AdamW, 
    get_linear_schedule_with_warmup,
    DataCollatorWithPadding,
    Trainer, 
    TrainingArguments,
    AutoModelForCausalLM
)

comet_ml is installed but `COMET_API_KEY` is not set.


In [ ]:
def print_number_of_trainable_model_parameters(model):
    trainable_model_params = 0
    all_model_params = 0
    for _, param in model.named_parameters():
        all_model_params += param.numel()
        if param.requires_grad:
            trainable_model_params += param.numel()
    return f"trainable model parameters: {trainable_model_params}\nall model parameters: {all_model_params}\n\
    percentage of trainable model parameters: {100 * trainable_model_params / all_model_params:.2f}%"
    

In [ ]:
data_file = "/home/scai/phd/aiz218323/scratch/datasets/train_v2_drcat_02.csv"

full_data = pd.read_csv(data_file)
full_data = full_data[full_data['RDizzl3_seven'] == True]
full_data = full_data[["text","label"]]
full_data.reset_index(drop=True, inplace = True)

print(f"We have {len(full_data)} samples") # Number of data we have
full_data.head(5)

We have 20450 samples


text  label
0  Cars have been around for awhile and they have...      0
1  Have you ever thought what it would be like no...      0
2  What you are about to read is going to give yo...      0
3  cars have many flaws nd and in this day and ag...      0
4  There are many advantages of limiting car usag...      0

In [ ]:
from sklearn.model_selection import train_test_split

# Split it when augmented data is ready
X_train, X_val, y_train, y_val = train_test_split(full_data["text"],
                                                  full_data["label"],
                                                  test_size=0.3,
                                                  stratify=full_data["label"],
                                                  random_state=42)
print(f"We have {len(X_train)} training samples")
print(f"We have {len(X_val)} validation samples")
print("----------------------------")
count = full_data["label"].value_counts()
print(f"Number of Essays written by Human: {count[0]}")
print(f"Number of Essays generated by LLM: {count[1]}")

X_train.reset_index(drop = True, inplace = True)
y_train.reset_index(drop = True, inplace = True)
X_val.reset_index(drop = True, inplace = True)
y_val.reset_index(drop = True, inplace = True)

We have 14315 training samples
We have 6135 validation samples
----------------------------
Number of Essays written by Human: 14250
Number of Essays generated by LLM: 6200


In [ ]:
from transformers import AutoModelForSequenceClassification,AutoTokenizer

In [ ]:
MODEL_PATH = 'distilbert/distilbert-base-uncased'
tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)

# Number of trainable parameters
print(print_number_of_trainable_model_parameters(model))

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert/distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


trainable model parameters: 66955010
all model parameters: 66955010
    percentage of trainable model parameters: 100.00%


In [ ]:
# Define the LoRA Configuration
lora_config = LoraConfig(
    r=8, # Rank Number
    lora_alpha=32, # Alpha (Scaling Factor)
    lora_dropout=0.05, # Dropout Prob for Lora
    target_modules=["q_lin", "k_lin","v_lin"], # Which layer to apply LoRA, usually only apply on MultiHead Attention Layer
    bias='none',
    task_type=TaskType.SEQ_CLS # Seqence to Classification Task
)

In [ ]:
def func():
    import pdb; pdb.set_trace()
    return get_peft_model(model, lora_config)
    

In [ ]:
model = AutoModelForSequenceClassification.from_pretrained(MODEL_PATH, return_dict=True, num_labels=2)
o = func()

Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert/distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier.bias', 'pre_classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


> /tmp/ipykernel_27181/4044093400.py(3)func()
      1 def func():
      2     import pdb; pdb.set_trace()
----> 3     return get_peft_model(model, lora_config)
      4 



ipdb>  model


DistilBertForSequenceClassification(
  (distilbert): DistilBertModel(
    (embeddings): Embeddings(
      (word_embeddings): Embedding(30522, 768, padding_idx=0)
      (position_embeddings): Embedding(512, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (transformer): Transformer(
      (layer): ModuleList(
        (0-5): 6 x TransformerBlock(
          (attention): MultiHeadSelfAttention(
            (dropout): Dropout(p=0.1, inplace=False)
            (q_lin): Linear(in_features=768, out_features=768, bias=True)
            (k_lin): Linear(in_features=768, out_features=768, bias=True)
            (v_lin): Linear(in_features=768, out_features=768, bias=True)
            (out_lin): Linear(in_features=768, out_features=768, bias=True)
          )
          (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (ffn): FFN(
            (dropout): Dropout(p=0.1, inplace=False)
 

ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(133)get_peft_model()
    131 
    132 
--> 133 def get_peft_model(
    134     model: PreTrainedModel,
    135     peft_config: PeftConfig,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(161)get_peft_model()
    159             the base model
    160     """
--> 161     model_config = getattr(model, "config", {"model_type": "custom"})
    162     if hasattr(model_config, "to_dict"):
    163         model_config = model_config.to_dict()



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(162)get_peft_model()
    160     """
    161     model_config = getattr(model, "config", {"model_type": "custom"})
--> 162     if hasattr(model_config, "to_dict"):
    163         model_config = model_config.to_dict()
    164 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(163)get_peft_model()
    161     model_config = getattr(model, "config", {"model_type": "custom"})
    162     if hasattr(model_config, "to_dict"):
--> 163         model_config = model_config.to_dict()
    164 
    165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(165)get_peft_model()
    163         model_config = model_config.to_dict()
    164 
--> 165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)
    166 
    167     if revision is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(167)get_peft_model()
    165     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)
    166 
--> 167     if revision is not None:
    168         if peft_config.revision is not None and peft_config.revision != revision:
    169             warnings.warn(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(174)get_peft_model()
    172         peft_config.revision = revision
    173 
--> 174     if mixed:
    175         # note: PeftMixedModel does not support autocast_adapter_dtype, so don't pass it
    176         return PeftMixedModel(model, peft_config, adapter_name=adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(178)get_peft_model()
    176         return PeftMixedModel(model, peft_config, adapter_name=adapter_name)
    177 
--> 178     if peft_config.task_type not in MODEL_TYPE_TO_PEFT_MODEL_MAPPING.keys() and not peft_config.is_prompt_learning:
    179         return PeftModel(model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype)
    180 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(181)get_peft_model()
    179         return PeftModel(model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype)
    180 
--> 181     if peft_config.is_prompt_learning:
    182         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
    183     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(183)get_peft_model()
    181     if peft_config.is_prompt_learning:
    182         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
--> 183     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](
    184         model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype
    185     )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(184)get_peft_model()
    182         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
    183     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](
--> 184         model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype
    185     )
    186 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/mapping.py(183)get_peft_model()
    181     if peft_config.is_prompt_learning:
    182         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
--> 183     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](
    184         model, peft_config, adapter_name=adapter_name, autocast_adapter_dtype=autocast_adapter_dtype
    185     )



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(1308)__init__()
   1306     """
   1307 
-> 1308     def __init__(
   1309         self, model: torch.nn.Module, peft_config: PeftConfig, adapter_name: str = "default", **kwargs
   1310     ) -> None:



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(1311)__init__()
   1309         self, model: torch.nn.Module, peft_config: PeftConfig, adapter_name: str = "default", **kwargs
   1310     ) -> None:
-> 1311         super().__init__(model, peft_config, adapter_name, **kwargs)
   1312 
   1313         classifier_module_names = ["classifier", "score"]



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(132)__init__()
    130     """
    131 
--> 132     def __init__(
    133         self,
    134         model: PreTrainedModel,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(139)__init__()
    137         autocast_adapter_dtype: bool = True,
    138     ) -> None:
--> 139         super().__init__()
    140         self.modules_to_save = None
    141         self.active_adapter = adapter_name



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(140)__init__()
    138     ) -> None:
    139         super().__init__()
--> 140         self.modules_to_save = None
    141         self.active_adapter = adapter_name
    142         self.peft_type = peft_config.peft_type



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(141)__init__()
    139         super().__init__()
    140         self.modules_to_save = None
--> 141         self.active_adapter = adapter_name
    142         self.peft_type = peft_config.peft_type
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(142)__init__()
    140         self.modules_to_save = None
    141         self.active_adapter = adapter_name
--> 142         self.peft_type = peft_config.peft_type
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to
    144         # forward.



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(145)__init__()
    143         # These args are special PEFT arguments that users can pass. They need to be removed before passing them to
    144         # forward.
--> 145         self.special_peft_forward_args = {"adapter_names"}
    146 
    147         self._is_prompt_learning = peft_config.is_prompt_learning



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(147)__init__()
    145         self.special_peft_forward_args = {"adapter_names"}
    146 
--> 147         self._is_prompt_learning = peft_config.is_prompt_learning
    148         if self._is_prompt_learning:
    149             self._peft_config = {adapter_name: peft_config}



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(148)__init__()
    146 
    147         self._is_prompt_learning = peft_config.is_prompt_learning
--> 148         if self._is_prompt_learning:
    149             self._peft_config = {adapter_name: peft_config}
    150             self.base_model = model



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(153)__init__()
    151             self.add_adapter(adapter_name, peft_config)
    152         else:
--> 153             self._peft_config = None
    154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
    155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(154)__init__()
    152         else:
    153             self._peft_config = None
--> 154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
    155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
    156             self.set_additional_trainable_modules(peft_config, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/peft_model.py(155)__init__()
    153             self._peft_config = None
    154             cls = PEFT_TYPE_TO_MODEL_MAPPING[peft_config.peft_type]
--> 155             self.base_model = cls(model, {adapter_name: peft_config}, adapter_name)
    156             self.set_additional_trainable_modules(peft_config, adapter_name)
    157 



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(138)__init__()
    136     prefix: str = "lora_"
    137 
--> 138     def __init__(self, model, config, adapter_name) -> None:
    139         super().__init__(model, config, adapter_name)
    140 



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(139)__init__()
    137 
    138     def __init__(self, model, config, adapter_name) -> None:
--> 139         super().__init__(model, config, adapter_name)
    140 
    141     def _check_new_adapter_config(self, config: LoraConfig) -> None:



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(146)__init__()
    144     """
    145 
--> 146     def __init__(
    147         self,
    148         model,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(152)__init__()
    150         adapter_name: str,
    151     ) -> None:
--> 152         super().__init__()
    153 
    154         self.model = model



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(154)__init__()
    152         super().__init__()
    153 
--> 154         self.model = model
    155         self.targeted_module_names: list[str] = []
    156 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(155)__init__()
    153 
    154         self.model = model
--> 155         self.targeted_module_names: list[str] = []
    156 
    157         # For advanced developers, if you want to attach multiple adapters to your



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(159)__init__()
    157         # For advanced developers, if you want to attach multiple adapters to your
    158         # model, just add a `peft_config` dict attribute to your model.
--> 159         if not hasattr(self, "peft_config"):
    160             self.peft_config = {adapter_name: peft_config} if isinstance(peft_config, PeftConfig) else peft_config
    161         else:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(160)__init__()
    158         # model, just add a `peft_config` dict attribute to your model.
    159         if not hasattr(self, "peft_config"):
--> 160             self.peft_config = {adapter_name: peft_config} if isinstance(peft_config, PeftConfig) else peft_config
    161         else:
    162             logger.info(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(172)__init__()
    170                 self.peft_config.update(peft_config)
    171 
--> 172         self.active_adapter: str | list[str] = adapter_name
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(173)__init__()
    171 
    172         self.active_adapter: str | list[str] = adapter_name
--> 173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
    175             self.inject_adapter(self.model, adapter_name)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(174)__init__()
    172         self.active_adapter: str | list[str] = adapter_name
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
--> 174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
    175             self.inject_adapter(self.model, adapter_name)
    176 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(175)__init__()
    173         self._pre_injection_hook(self.model, self.peft_config[adapter_name], adapter_name)
    174         if peft_config != PeftType.XLORA or peft_config[adapter_name] != PeftType.XLORA:
--> 175             self.inject_adapter(self.model, adapter_name)
    176 
    177         # Copy the peft_config in the injected model.



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(353)__getattr__()
    351         return new_module
    352 
--> 353     def __getattr__(self, name: str):
    354         """Forward missing attributes to the wrapped module."""
    355         try:



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(355)__getattr__()
    353     def __getattr__(self, name: str):
    354         """Forward missing attributes to the wrapped module."""
--> 355         try:
    356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(356)__getattr__()
    354         """Forward missing attributes to the wrapped module."""
    355         try:
--> 356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:
    358             if name == "model":  # see #1892: prevent infinite recursion if class is not initialized



ipdb>  


--Return--
DistilBertFor...place=False)
)
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(356)__getattr__()
    354         """Forward missing attributes to the wrapped module."""
    355         try:
--> 356             return super().__getattr__(name)  # defer to nn.Module's logic
    357         except AttributeError:
    358             if name == "model":  # see #1892: prevent infinite recursion if class is not initialized



ipdb>  


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(366)inject_adapter()
    364         pass
    365 
--> 366     def inject_adapter(self, model: nn.Module, adapter_name: str, autocast_adapter_dtype: bool = True) -> None:
    367         r"""
    368         Creates adapter layers and replaces the target modules with the adapter layers. This method is called under the



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(381)inject_adapter()
    379                 Whether to autocast the adapter dtype. Defaults to `True`.
    380         """
--> 381         peft_config = self.peft_config[adapter_name]
    382         # Note: If possible, all checks should be performed *at the start of this method*.
    383         # This way, we can raise early if something goes wrong, without leaving the model



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(385)inject_adapter()
    383         # This way, we can raise early if something goes wrong, without leaving the model
    384         # in a bad (half-initialized) state.
--> 385         self._check_new_adapter_config(peft_config)
    386 
    387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(387)inject_adapter()
    385         self._check_new_adapter_config(peft_config)
    386 
--> 387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None
    388         _has_modules_to_save = False
    389 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(388)inject_adapter()
    386 
    387         _check_for_modules_to_save = getattr(peft_config, "modules_to_save", None) is not None
--> 388         _has_modules_to_save = False
    389 
    390         model_config = getattr(model, "config", {"model_type": "custom"})



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(390)inject_adapter()
    388         _has_modules_to_save = False
    389 
--> 390         model_config = getattr(model, "config", {"model_type": "custom"})
    391         if hasattr(model_config, "to_dict"):
    392             model_config = model_config.to_dict()



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(391)inject_adapter()
    389 
    390         model_config = getattr(model, "config", {"model_type": "custom"})
--> 391         if hasattr(model_config, "to_dict"):
    392             model_config = model_config.to_dict()
    393 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(392)inject_adapter()
    390         model_config = getattr(model, "config", {"model_type": "custom"})
    391         if hasattr(model_config, "to_dict"):
--> 392             model_config = model_config.to_dict()
    393 
    394         peft_config = self._prepare_adapter_config(peft_config, model_config)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(394)inject_adapter()
    392             model_config = model_config.to_dict()
    393 
--> 394         peft_config = self._prepare_adapter_config(peft_config, model_config)
    395 
    396         self._prepare_model(peft_config, model)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(396)inject_adapter()
    394         peft_config = self._prepare_adapter_config(peft_config, model_config)
    395 
--> 396         self._prepare_model(peft_config, model)
    397         is_target_modules_in_base_model = False
    398         key_list = [key for key, _ in model.named_modules()]



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(397)inject_adapter()
    395 
    396         self._prepare_model(peft_config, model)
--> 397         is_target_modules_in_base_model = False
    398         key_list = [key for key, _ in model.named_modules()]
    399 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(398)inject_adapter()
    396         self._prepare_model(peft_config, model)
    397         is_target_modules_in_base_model = False
--> 398         key_list = [key for key, _ in model.named_modules()]
    399 
    400         if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(400)inject_adapter()
    398         key_list = [key for key, _ in model.named_modules()]
    399 
--> 400         if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:
    401             # dummy adapter, we allow not matching any module
    402             key_list = []



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(406)inject_adapter()
    404 
    405         # update peft_config.target_modules if required
--> 406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
    408         for key in key_list:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(408)inject_adapter()
    406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
--> 408         for key in key_list:
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(411)inject_adapter()
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(
--> 411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):
    413                 # Optionally set the modules to save



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(425)inject_adapter()
    423                 continue
    424 
--> 425             if not self._check_target_module_exists(peft_config, key):
    426                 continue
    427 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(426)inject_adapter()
    424 
    425             if not self._check_target_module_exists(peft_config, key):
--> 426                 continue
    427 
    428             self.targeted_module_names.append(key)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(408)inject_adapter()
    406         peft_config = _maybe_include_all_linear_layers(peft_config, model)
    407 
--> 408         for key in key_list:
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(411)inject_adapter()
    409             # Check for modules_to_save in case
    410             if _check_for_modules_to_save and any(
--> 411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):
    413                 # Optionally set the modules to save



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(410)inject_adapter()
    408         for key in key_list:
    409             # Check for modules_to_save in case
--> 410             if _check_for_modules_to_save and any(
    411                 key.endswith(f"{module_to_save}") for module_to_save in peft_config.modules_to_save
    412             ):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(425)inject_adapter()
    423                 continue
    424 
--> 425             if not self._check_target_module_exists(peft_config, key):
    426                 continue
    427 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(426)inject_adapter()
    424 
    425             if not self._check_target_module_exists(peft_config, key):
--> 426                 continue
    427 
    428             self.targeted_module_names.append(key)



ipdb>  b 428


Breakpoint 6 at /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py:428


ipdb>  r


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(428)inject_adapter()
    426                 continue
    427 
6-> 428             self.targeted_module_names.append(key)
    429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)



ipdb>  type(model)


<class 'transformers.models.distilbert.modeling_distilbert.DistilBertForSequenceClassification'>


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(429)inject_adapter()
    427 
6   428             self.targeted_module_names.append(key)
--> 429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)
    431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(430)inject_adapter()
6   428             self.targeted_module_names.append(key)
    429             is_target_modules_in_base_model = True
--> 430             parent, target, target_name = _get_submodules(model, key)
    431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
    432 



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(310)_get_submodules()
    308 
    309 
--> 310 def _get_submodules(model, key):
    311     parent = model.get_submodule(".".join(key.split(".")[:-1]))
    312     target_name = key.split(".")[-1]



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(311)_get_submodules()
    309 
    310 def _get_submodules(model, key):
--> 311     parent = model.get_submodule(".".join(key.split(".")[:-1]))
    312     target_name = key.split(".")[-1]
    313     target = model.get_submodule(key)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(312)_get_submodules()
    310 def _get_submodules(model, key):
    311     parent = model.get_submodule(".".join(key.split(".")[:-1]))
--> 312     target_name = key.split(".")[-1]
    313     target = model.get_submodule(key)
    314     return parent, target, target_name



ipdb>  parent


MultiHeadSelfAttention(
  (dropout): Dropout(p=0.1, inplace=False)
  (q_lin): Linear(in_features=768, out_features=768, bias=True)
  (k_lin): Linear(in_features=768, out_features=768, bias=True)
  (v_lin): Linear(in_features=768, out_features=768, bias=True)
  (out_lin): Linear(in_features=768, out_features=768, bias=True)
)


ipdb>  key


'distilbert.transformer.layer.0.attention.q_lin'


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(313)_get_submodules()
    311     parent = model.get_submodule(".".join(key.split(".")[:-1]))
    312     target_name = key.split(".")[-1]
--> 313     target = model.get_submodule(key)
    314     return parent, target, target_name
    315 



ipdb>  target_name


'q_lin'


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(314)_get_submodules()
    312     target_name = key.split(".")[-1]
    313     target = model.get_submodule(key)
--> 314     return parent, target, target_name
    315 
    316 



ipdb>  target


Linear(in_features=768, out_features=768, bias=True)


ipdb>  n


--Return--
(MultiHeadSelf..., bias=True)
), Linear(in_fea...68, bias=True), 'q_lin')
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/utils/other.py(314)_get_submodules()
    312     target_name = key.split(".")[-1]
    313     target = model.get_submodule(key)
--> 314     return parent, target, target_name
    315 
    316 



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/tuners_utils.py(431)inject_adapter()
    429             is_target_modules_in_base_model = True
    430             parent, target, target_name = _get_submodules(model, key)
--> 431             self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
    432 
    433         # Handle X-LoRA case.



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(173)_create_and_replace()
    171             replicate_layers(model, peft_config.layer_replication)
    172 
--> 173     def _create_and_replace(
    174         self,
    175         lora_config,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(182)_create_and_replace()
    180         current_key,
    181     ):
--> 182         if current_key is None:
    183             raise ValueError("Current Key shouldn't be `None`")
    184 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(186)_create_and_replace()
    184 
    185         # Regexp matching - Find key which matches current target_name in patterns provided
--> 186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(187)_create_and_replace()
    185         # Regexp matching - Find key which matches current target_name in patterns provided
    186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
--> 187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)



ipdb>  pattern_keys


[]


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(188)_create_and_replace()
    186         pattern_keys = list(chain(lora_config.rank_pattern.keys(), lora_config.alpha_pattern.keys()))
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
--> 188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 



ipdb>  target_name_key


'distilbert.transformer.layer.0.attention.q_lin'


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(189)_create_and_replace()
    187         target_name_key = next(filter(lambda key: re.match(rf".*\.{key}$", current_key), pattern_keys), current_key)
    188         r = lora_config.rank_pattern.get(target_name_key, lora_config.r)
--> 189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 
    191         kwargs = {



ipdb>  !r


8


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(192)_create_and_replace()
    190 
    191         kwargs = {
--> 192             "r": r,
    193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(193)_create_and_replace()
    191         kwargs = {
    192             "r": r,
--> 193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(194)_create_and_replace()
    192             "r": r,
    193             "lora_alpha": alpha,
--> 194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(195)_create_and_replace()
    193             "lora_alpha": alpha,
    194             "lora_dropout": lora_config.lora_dropout,
--> 195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(196)_create_and_replace()
    194             "lora_dropout": lora_config.lora_dropout,
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
--> 196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(197)_create_and_replace()
    195             "fan_in_fan_out": lora_config.fan_in_fan_out,
    196             "init_lora_weights": lora_config.init_lora_weights,
--> 197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(198)_create_and_replace()
    196             "init_lora_weights": lora_config.init_lora_weights,
    197             "use_rslora": lora_config.use_rslora,
--> 198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(199)_create_and_replace()
    197             "use_rslora": lora_config.use_rslora,
    198             "use_dora": lora_config.use_dora,
--> 199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
    201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(200)_create_and_replace()
    198             "use_dora": lora_config.use_dora,
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
--> 200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
    201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),
    202         }



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(201)_create_and_replace()
    199             "ephemeral_gpu_offload": lora_config.runtime_config.ephemeral_gpu_offload,
    200             "loaded_in_8bit": getattr(self.model, "is_loaded_in_8bit", False),
--> 201             "loaded_in_4bit": getattr(self.model, "is_loaded_in_4bit", False),
    202         }
    203 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(191)_create_and_replace()
    189         alpha = lora_config.alpha_pattern.get(target_name_key, lora_config.lora_alpha)
    190 
--> 191         kwargs = {
    192             "r": r,
    193             "lora_alpha": alpha,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(204)_create_and_replace()
    202         }
    203 
--> 204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(206)_create_and_replace()
    204         quant_methods = ["gptq", "aqlm", "awq"]
    205         for quant_method in quant_methods:
--> 206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(207)_create_and_replace()
    205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
--> 207             if quantization_config is not None:
    208                 kwargs[f"{quant_method}_quantization_config"] = quantization_config
    209 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(205)_create_and_replace()
    203 
    204         quant_methods = ["gptq", "aqlm", "awq"]
--> 205         for quant_method in quant_methods:
    206             quantization_config = get_quantization_config(self.model, method=quant_method)
    207             if quantization_config is not None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(211)_create_and_replace()
    209 
    210         # note: AdaLoraLayer is a subclass of LoraLayer, we need to exclude it
--> 211         from peft.tuners.adalora import AdaLoraLayer
    212 
    213         if isinstance(target, LoraLayer) and not isinstance(target, AdaLoraLayer):



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(213)_create_and_replace()
    211         from peft.tuners.adalora import AdaLoraLayer
    212 
--> 213         if isinstance(target, LoraLayer) and not isinstance(target, AdaLoraLayer):
    214             target.update_layer(
    215                 adapter_name,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(224)_create_and_replace()
    222             )
    223         else:
--> 224             new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
    225             if adapter_name not in self.active_adapters:
    226                 # adding an additional adapter: it is not automatically trainable



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(289)_create_new_module()
    287                 raise NotImplementedError(f"Requested bias: {bias}, is not implemented.")
    288 
--> 289     @staticmethod
    290     def _create_new_module(lora_config, adapter_name, target, **kwargs):
    291         # Collect dispatcher functions to decide what backend to use for the replaced LoRA layer. The order matters,



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(293)_create_new_module()
    291         # Collect dispatcher functions to decide what backend to use for the replaced LoRA layer. The order matters,
    292         # because the first match is always used. Therefore, the default layers should be checked last.
--> 293         dispatchers = []
    294 
    295         if lora_config._custom_modules:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(295)_create_new_module()
    293         dispatchers = []
    294 
--> 295         if lora_config._custom_modules:
    296             # Experimental custom LoRA module support. Allows users to pass a custom mapping for unsupported layer
    297             # types by impelementing their own LoRA layers.



ipdb>  lora_config._custom_modules
ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(316)_create_new_module()
    314 
    315         # avoid eager bnb import
--> 316         if is_bnb_available():
    317             from .bnb import dispatch_bnb_8bit
    318 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(321)_create_new_module()
    319             dispatchers.append(dispatch_bnb_8bit)
    320 
--> 321         if is_bnb_4bit_available():
    322             from .bnb import dispatch_bnb_4bit
    323 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(326)_create_new_module()
    324             dispatchers.append(dispatch_bnb_4bit)
    325 
--> 326         dispatchers.extend(
    327             [
    328                 dispatch_eetq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(328)_create_new_module()
    326         dispatchers.extend(
    327             [
--> 328                 dispatch_eetq,
    329                 dispatch_aqlm,
    330                 dispatch_awq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(329)_create_new_module()
    327             [
    328                 dispatch_eetq,
--> 329                 dispatch_aqlm,
    330                 dispatch_awq,
    331                 dispatch_gptq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(330)_create_new_module()
    328                 dispatch_eetq,
    329                 dispatch_aqlm,
--> 330                 dispatch_awq,
    331                 dispatch_gptq,
    332                 dispatch_hqq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(331)_create_new_module()
    329                 dispatch_aqlm,
    330                 dispatch_awq,
--> 331                 dispatch_gptq,
    332                 dispatch_hqq,
    333                 dispatch_megatron,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(332)_create_new_module()
    330                 dispatch_awq,
    331                 dispatch_gptq,
--> 332                 dispatch_hqq,
    333                 dispatch_megatron,
    334                 dispatch_default,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(333)_create_new_module()
    331                 dispatch_gptq,
    332                 dispatch_hqq,
--> 333                 dispatch_megatron,
    334                 dispatch_default,
    335             ]



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(334)_create_new_module()
    332                 dispatch_hqq,
    333                 dispatch_megatron,
--> 334                 dispatch_default,
    335             ]
    336         )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(327)_create_new_module()
    325 
    326         dispatchers.extend(
--> 327             [
    328                 dispatch_eetq,
    329                 dispatch_aqlm,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(326)_create_new_module()
    324             dispatchers.append(dispatch_bnb_4bit)
    325 
--> 326         dispatchers.extend(
    327             [
    328                 dispatch_eetq,



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(338)_create_new_module()
    336         )
    337 
--> 338         new_module = None
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  dispatcher


<function dispatch_eetq>


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(339)_create_new_module()
    337 
    338         new_module = None
--> 339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(340)_create_new_module()
    338         new_module = None
    339         for dispatcher in dispatchers:
--> 340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
    342                 break



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(341)_create_new_module()
    339         for dispatcher in dispatchers:
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
--> 341             if new_module is not None:  # first match wins
    342                 break
    343 



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(342)_create_new_module()
    340             new_module = dispatcher(target, adapter_name, lora_config=lora_config, **kwargs)
    341             if new_module is not None:  # first match wins
--> 342                 break
    343 
    344         if new_module is None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(344)_create_new_module()
    342                 break
    343 
--> 344         if new_module is None:
    345             # no module could be matched
    346             raise ValueError(



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(351)_create_new_module()
    349             )
    350 
--> 351         return new_module
    352 
    353     def __getattr__(self, name: str):



ipdb>  new_module


lora.Linear(
  (base_layer): Linear(in_features=768, out_features=768, bias=True)
  (lora_dropout): ModuleDict(
    (default): Dropout(p=0.05, inplace=False)
  )
  (lora_A): ModuleDict(
    (default): Linear(in_features=768, out_features=8, bias=False)
  )
  (lora_B): ModuleDict(
    (default): Linear(in_features=8, out_features=768, bias=False)
  )
  (lora_embedding_A): ParameterDict()
  (lora_embedding_B): ParameterDict()
  (lora_magnitude_vector): ModuleDict()
)


ipdb>  n


--Return--
lora.Linear(
...ModuleDict()
)
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(351)_create_new_module()
    349             )
    350 
--> 351         return new_module
    352 
    353     def __getattr__(self, name: str):



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(225)_create_and_replace()
    223         else:
    224             new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
--> 225             if adapter_name not in self.active_adapters:
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)



ipdb>  !adapter_name


'default'


ipdb>  !self.active_adapters


['default']


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(228)_create_and_replace()
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)
--> 228             self._replace_module(parent, target_name, new_module, target)
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):



ipdb>  parent


MultiHeadSelfAttention(
  (dropout): Dropout(p=0.1, inplace=False)
  (q_lin): Linear(in_features=768, out_features=768, bias=True)
  (k_lin): Linear(in_features=768, out_features=768, bias=True)
  (v_lin): Linear(in_features=768, out_features=768, bias=True)
  (out_lin): Linear(in_features=768, out_features=768, bias=True)
)


ipdb>  target_name


'q_lin'


ipdb>  new_module


lora.Linear(
  (base_layer): Linear(in_features=768, out_features=768, bias=True)
  (lora_dropout): ModuleDict(
    (default): Dropout(p=0.05, inplace=False)
  )
  (lora_A): ModuleDict(
    (default): Linear(in_features=768, out_features=8, bias=False)
  )
  (lora_B): ModuleDict(
    (default): Linear(in_features=8, out_features=768, bias=False)
  )
  (lora_embedding_A): ParameterDict()
  (lora_embedding_B): ParameterDict()
  (lora_magnitude_vector): ModuleDict()
)


ipdb>  target


Linear(in_features=768, out_features=768, bias=True)


ipdb>  l


    223         else:
    224             new_module = self._create_new_module(lora_config, adapter_name, target, **kwargs)
    225             if adapter_name not in self.active_adapters:
    226                 # adding an additional adapter: it is not automatically trainable
    227                 new_module.requires_grad_(False)
--> 228             self._replace_module(parent, target_name, new_module, target)
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):
    231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by
    233         # _mark_only_adapters_as_trainable



ipdb>  s


--Call--
> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(230)_replace_module()
    228             self._replace_module(parent, target_name, new_module, target)
    229 
--> 230     def _replace_module(self, parent, child_name, new_module, child):
    231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(231)_replace_module()
    229 
    230     def _replace_module(self, parent, child_name, new_module, child):
--> 231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by
    233         # _mark_only_adapters_as_trainable



ipdb>  parent


MultiHeadSelfAttention(
  (dropout): Dropout(p=0.1, inplace=False)
  (q_lin): Linear(in_features=768, out_features=768, bias=True)
  (k_lin): Linear(in_features=768, out_features=768, bias=True)
  (v_lin): Linear(in_features=768, out_features=768, bias=True)
  (out_lin): Linear(in_features=768, out_features=768, bias=True)
)


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(236)_replace_module()
    234 
    235         # child layer wraps the original module, unpack it
--> 236         if hasattr(child, "base_layer"):
    237             child = child.base_layer
    238 



ipdb>  parent


MultiHeadSelfAttention(
  (dropout): Dropout(p=0.1, inplace=False)
  (q_lin): lora.Linear(
    (base_layer): Linear(in_features=768, out_features=768, bias=True)
    (lora_dropout): ModuleDict(
      (default): Dropout(p=0.05, inplace=False)
    )
    (lora_A): ModuleDict(
      (default): Linear(in_features=768, out_features=8, bias=False)
    )
    (lora_B): ModuleDict(
      (default): Linear(in_features=8, out_features=768, bias=False)
    )
    (lora_embedding_A): ParameterDict()
    (lora_embedding_B): ParameterDict()
    (lora_magnitude_vector): ModuleDict()
  )
  (k_lin): Linear(in_features=768, out_features=768, bias=True)
  (v_lin): Linear(in_features=768, out_features=768, bias=True)
  (out_lin): Linear(in_features=768, out_features=768, bias=True)
)


ipdb>  l


    231         setattr(parent, child_name, new_module)
    232         # It's not necessary to set requires_grad here, as that is handled by
    233         # _mark_only_adapters_as_trainable
    234 
    235         # child layer wraps the original module, unpack it
--> 236         if hasattr(child, "base_layer"):
    237             child = child.base_layer
    238 
    239         if not hasattr(new_module, "base_layer"):
    240             if hasattr(new_module, "W_q"):  # HQQ
    241                 new_module.W_q = child.W_q



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(239)_replace_module()
    237             child = child.base_layer
    238 
--> 239         if not hasattr(new_module, "base_layer"):
    240             if hasattr(new_module, "W_q"):  # HQQ
    241                 new_module.W_q = child.W_q



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(247)_replace_module()
    245                 new_module.bias = child.bias
    246 
--> 247         if getattr(child, "state", None) is not None:
    248             if hasattr(new_module, "base_layer"):
    249                 new_module.base_layer.state = child.state



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  self.prefix


'lora_'


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(256)_replace_module()
    254         # dispatch to correct device
    255         for name, module in new_module.named_modules():
--> 256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
    258                     child.qweight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(259)_replace_module()
    257                 weight = (
    258                     child.qweight
--> 259                     if hasattr(child, "qweight")
    260                     else child.W_q
    261                     if hasattr(child, "W_q")



ipdb>  module


ModuleDict(
  (default): Dropout(p=0.05, inplace=False)
)


ipdb>  n


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(258)_replace_module()
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (
--> 258                     child.qweight
    259                     if hasattr(child, "qweight")
    260                     else child.W_q



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(261)_replace_module()
    259                     if hasattr(child, "qweight")
    260                     else child.W_q
--> 261                     if hasattr(child, "W_q")
    262                     else child.weight
    263                     if hasattr(child, "weight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(260)_replace_module()
    258                     child.qweight
    259                     if hasattr(child, "qweight")
--> 260                     else child.W_q
    261                     if hasattr(child, "W_q")
    262                     else child.weight



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(263)_replace_module()
    261                     if hasattr(child, "W_q")
    262                     else child.weight
--> 263                     if hasattr(child, "weight")
    264                     else next(child.parameters())
    265                 )



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(262)_replace_module()
    260                     else child.W_q
    261                     if hasattr(child, "W_q")
--> 262                     else child.weight
    263                     if hasattr(child, "weight")
    264                     else next(child.parameters())



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(257)_replace_module()
    255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
--> 257                 weight = (
    258                     child.qweight
    259                     if hasattr(child, "qweight")



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(266)_replace_module()
    264                     else next(child.parameters())
    265                 )
--> 266                 module.to(weight.device)
    267 
    268     def _mark_only_adapters_as_trainable(self, model: nn.Module) -> None:



ipdb>  


> /scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(255)_replace_module()
    253 
    254         # dispatch to correct device
--> 255         for name, module in new_module.named_modules():
    256             if (self.prefix in name) or ("ranknum" in name):
    257                 weight = (



ipdb>  new_module


lora.Linear(
  (base_layer): Linear(in_features=768, out_features=768, bias=True)
  (lora_dropout): ModuleDict(
    (default): Dropout(p=0.05, inplace=False)
  )
  (lora_A): ModuleDict(
    (default): Linear(in_features=768, out_features=8, bias=False)
  )
  (lora_B): ModuleDict(
    (default): Linear(in_features=8, out_features=768, bias=False)
  )
  (lora_embedding_A): ParameterDict()
  (lora_embedding_B): ParameterDict()
  (lora_magnitude_vector): ModuleDict()
)


ipdb>  q


/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(224)
/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg_2/lib/python3.9/site-packages/peft/tuners/lora/model.py(211)


In [ ]:
# Get our LoRA-enabled model
peft_model = get_peft_model(model, lora_config)

# Reduced trainble parameters
print(print_number_of_trainable_model_parameters(peft_model))

trainable model parameters: 813314
all model parameters: 67768324
    percentage of trainable model parameters: 1.20%


In [ ]:
import datasets

In [ ]:
def tokenize_func(data):
    return tokenizer(
            data['texts'],
            max_length=512,
            padding='max_length',
            return_attention_mask=True,
            truncation=True
        )

In [ ]:
train_dataset = datasets.Dataset.from_pandas(pd.DataFrame({"texts":X_train,"labels":y_train}))
train_dataset = train_dataset.map(
    tokenize_func,
    batched=True,
    remove_columns=["texts"]
)
train_dataset

Map:   0%|          | 0/14315 [00:00<?, ? examples/s]

Dataset({
    features: ['labels', 'input_ids', 'attention_mask'],
    num_rows: 14315
})

In [ ]:
# Tokenize the Validation Data
val_dataset = datasets.Dataset.from_pandas(pd.DataFrame({"texts":X_val,"labels":y_val}))
val_dataset = val_dataset.map(
    tokenize_func,
    batched=True,
    remove_columns=["texts"]
)

val_dataset

Map:   0%|          | 0/6135 [00:00<?, ? examples/s]

Dataset({
    features: ['labels', 'input_ids', 'attention_mask'],
    num_rows: 6135
})

In [ ]:
def metrics(eval_prediction):
    logits, labels = eval_prediction
    pred = np.argmax(logits, axis=1)
    auc_score = roc_auc_score(labels, pred)
    return {"Val-AUC": auc_score}

train_batch_size = 32
eval_batch_size = 32

In [ ]:
# Define training Args
peft_training_args = TrainingArguments(
    output_dir='/home/scai/phd/aiz218323/scratch/outputs/89-lora-distilbert',
    learning_rate=1e-4,
    per_device_train_batch_size=train_batch_size, # You can adjust this value base on your available GPU, You may encounter "out of memory" error if this value is too lartge
    per_device_eval_batch_size=eval_batch_size, # You can adjust this value base on your available GPU, You may encounter "out of memory" error if this value is too lartge
    num_train_epochs=5,
    logging_steps=10,
    evaluation_strategy='steps',
    eval_steps=10,
    weight_decay=0.01,
    seed=42,
    fp16=True, # Only use with GPU
    report_to='none'
)   

In [ ]:
import math
from sklearn.metrics import roc_auc_score

In [ ]:
# Define Optimzer
optimizer = AdamW(peft_model.parameters(), 
                  lr=1e-4,
                  no_deprecation_warning=True)

# Define Scheduler
n_epochs = peft_training_args.num_train_epochs
total_steps = n_epochs * math.ceil(len(train_dataset) / train_batch_size / 2)
lr_scheduler = get_linear_schedule_with_warmup(
    optimizer, 
    num_warmup_steps=0, 
    num_training_steps=total_steps)

# Data Collator
collator = DataCollatorWithPadding(
    tokenizer=tokenizer, 
    padding="longest"
)

# Define Trainer
peft_trainer = Trainer(
    model=peft_model,
    args=peft_training_args,
    train_dataset=train_dataset, # Training Data
    eval_dataset=val_dataset, # Evaluation Data
    tokenizer=tokenizer,
    compute_metrics=metrics,
    optimizers=(optimizer,lr_scheduler),
    data_collator=collator
)

print(f"Total Steps: {total_steps}")

/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/accelerate/accelerator.py:432: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
Detected kernel version 3.10.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.


Total Steps: 1120


In [ ]:
peft_trainer.train()

/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/site-packages/torch/nn/parallel/_functions.py:68: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/scratch/scai/phd/aiz218323/anaconda3/envs/xc_nlg/lib/python3.9/

TrainOutput(global_step=1120, training_loss=0.04091595844513444, metrics={'train_runtime': 3157.1177, 'train_samples_per_second': 22.671, 'train_steps_per_second': 0.355, 'total_flos': 9660184239820800.0, 'train_loss': 0.04091595844513444, 'epoch': 5.0})